File: app/src/main/java/com/mx/gillustrated/activity/GameListActivity.java
package com.mx.gillustrated.activity;

import java.util.ArrayList;
import java.util.List;

import com.j256.ormlite.dao.Dao;
import com.mx.gillustrated.R;
import com.mx.gillustrated.adapter.GameListAdapter;
import com.mx.gillustrated.adapter.GameListAdapter.DespairTouchListener;
import com.mx.gillustrated.listener.ListenerListViewScrollHandler;
import com.mx.gillustrated.vo.GameInfo;
import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.view.View;
import android.view.Window;
import android.widget.Button;
import android.widget.ImageButton;
import android.widget.ListView;
import android.widget.RelativeLayout;
import android.widget.Toast;

public class GameListActivity extends BaseActivity {
	
	private ImageButton mBtnAdd;
	private ListView mLvDespairMain;
	private GameListAdapter mAdapter;
	private List<GameInfo> mList;
	private RelativeLayout pageVboxLayout;
	
	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		requestWindowFeature(Window.FEATURE_NO_TITLE);
		setContentView(R.layout.activity_game);
			
		mBtnAdd = (ImageButton) findViewById(R.id.btnDespairAdd);
		mBtnAdd.setOnClickListener(onAddBtnClickListerner);
		
		mLvDespairMain = (ListView) findViewById(R.id.lvDespairMain);
		pageVboxLayout = (RelativeLayout) findViewById(R.id.pageVBox);
		pageVboxLayout.setVisibility(View.GONE);
		
		mLvDespairMain.setOnScrollListener(new ListenerListViewScrollHandler(mLvDespairMain, pageVboxLayout));
		mList = new ArrayList<GameInfo>();
		mAdapter = new GameListAdapter(this, mList);
		mAdapter.setDespairTouchListener(despairTouchListener);
		
		searchMain();
		
	}
	
	DespairTouchListener despairTouchListener = new DespairTouchListener(){

		@Override
		public void onSaveBtnClickListener(GameInfo info) {
			Dao.CreateOrUpdateStatus result = mOrmHelper.getGameInfoDao().createOrUpdate(info);
			if( result.isCreated() || result.isUpdated() ) {
				Toast.makeText(GameListActivity.this, result.isCreated() ? "新增成功" : "更新成功", Toast.LENGTH_SHORT).show();
				searchMain();
			}
				
		}

		@Override
		public void onDetailBtnClickListener(GameInfo info) {
			// TODO Auto-generated method stub
			Intent intent = new Intent(GameListActivity.this, GameInfoActivity.class);
			intent.putExtra("game", info.getId());
			startActivity(intent);
		}
		
	};
	
	View.OnClickListener onAddBtnClickListerner = new View.OnClickListener() {
		
		@Override
		public void onClick(View v) {
			mList.add(0, new GameInfo());
			updateList(false);
		}
	};
	
	Handler mainHandler = new Handler() {
		@Override
		public void handleMessage(Message msg) {
			if(msg.what == 1){
				mList.clear();
				mList.addAll((List<GameInfo>) msg.obj);
				updateList(true);
			}
		}
		
	};
	
	private void searchMain(){
		mainHandler.post( new Runnable() {
			@Override
			public void run() {
				List<GameInfo> list = mOrmHelper.getGameInfoDao().queryForAll();
				Message msg = mainHandler.obtainMessage();
				msg.what = 1;
				msg.obj = list;
				mainHandler.sendMessage(msg);
			}
		});
	}
	
	private void updateList(boolean flag) {
		if (flag)
			mLvDespairMain.setAdapter(mAdapter);
		else
			mAdapter.notifyDataSetChanged();
	}
	
}

File: app/src/main/java/com/mx/gillustrated/activity/GameListActivity.kt
package com.mx.gillustrated.activity

import com.mx.gillustrated.adapter.GameListAdapter
import com.mx.gillustrated.adapter.GameListAdapter.DespairTouchListener
import com.mx.gillustrated.listener.ListenerListViewScrollHandler
import com.mx.gillustrated.vo.GameInfo
import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.os.Message
import android.view.View
import android.view.Window
import android.widget.Toast
import com.mx.gillustrated.databinding.ActivityGameBinding
import java.lang.ref.WeakReference

class GameListActivity : BaseActivity() {

    private var mAdapter: GameListAdapter? = null
    private var mList: MutableList<GameInfo> = mutableListOf()
    private var mainHandler: MainHandler = MainHandler(this)
    lateinit var binding:ActivityGameBinding

    companion object {

        @Suppress("UNCHECKED_CAST")
        internal class MainHandler(activity: GameListActivity) : Handler(Looper.getMainLooper()){

            private val weakReference:WeakReference<GameListActivity> = WeakReference(activity)

            override fun handleMessage(msg: Message) {
                val activity = weakReference.get()!!
                if (msg.what == 1) {
                    activity.mList.clear()
                    activity.mList.addAll(msg.obj as List<GameInfo>)
                    activity.updateList(true)
                }
            }
        }

    }


    private var despairTouchListener: DespairTouchListener = object : DespairTouchListener {

        override fun onSaveBtnClickListener(info: GameInfo) {
            val result = mOrmHelper.gameInfoDao.createOrUpdate(info)
            if (result.isCreated || result.isUpdated) {
                Toast.makeText(this@GameListActivity, if (result.isCreated) "新增成功" else "更新成功", Toast.LENGTH_SHORT).show()
                searchMain()
            }

        }

        override fun onDetailBtnClickListener(info: GameInfo) {
            val intent = Intent(this@GameListActivity, GameInfoActivity::class.java)
            intent.putExtra("game", info.id)
            startActivity(intent)
        }

    }

    private var onAddBtnClickListerner: View.OnClickListener = View.OnClickListener {
        mList.add(0, GameInfo())
        updateList(false)
    }


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        requestWindowFeature(Window.FEATURE_NO_TITLE)
        binding = ActivityGameBinding.inflate(layoutInflater)
        setContentView(binding.root)


        binding.btnDespairAdd.setOnClickListener(onAddBtnClickListerner)
        binding.pageVBox.visibility = View.GONE
        binding.lvDespairMain.setOnScrollListener(ListenerListViewScrollHandler(binding.lvDespairMain, binding.pageVBox))
        mAdapter = GameListAdapter(this, mList)
        mAdapter!!.setDespairTouchListener(despairTouchListener)

        searchMain()

    }

    private fun searchMain() {
        mainHandler.post {
            val list = mOrmHelper.gameInfoDao.queryForAll()
            val msg = mainHandler.obtainMessage()
            msg.what = 1
            msg.obj = list
            mainHandler.sendMessage(msg)
        }
    }

    private fun updateList(flag: Boolean) {
        if (flag)
            binding.lvDespairMain.adapter = mAdapter
        else
            mAdapter!!.notifyDataSetChanged()
    }

}

